bpi-r64: add spidev for r64

PolynomialDivision opened the pull request:
Add spidev.

PolynomialDivision commented:
@dangowrt Run-tested on R2/R64.

dangowrt commented:
Wouldn't it be better to define a DT-overlay which contains what ever component you have added to the SPI bus?
Because other people may use it for different things on the bus and most of them will need a kernel driver rather than spi-dev in userspace. That would also allow to drop `910-add-spidev-to-spidev-dt-ids.patch` which won't ever be acceptable for upstream but rather define a meaningful compatible-string for the actual hardware sitting on the bus instead.

PolynomialDivision commented:
> Wouldn't it be better to define a DT-overlay which contains what ever component you have added to the SPI bus?

Great idea. Is there some example already existing for OpenWrt? I read the [kernel documentation](https://www.kernel.org/doc/Documentation/devicetree/overlay-notes.txt).

PolynomialDivision commented:
Just going through OpenWrt code and realizing that there is already DTS_OVERLAY. Do we talk about changing the dts at compile or runtime?

dangowrt commented:
We are already using DT overlay to select either SATA or mPCIe configuration for the R64.
It works in such way that the overlay can be part of the uImage.FIT and U-Boot patches the tree at runtime depending on the selected configuration.
In your case you could as well use a locally stored DT overlay (eg. at an offset at the end of `recovery` partition) for your SPI setup.

PolynomialDivision committed:
mediatek: bpi-r64: add qorvo-dw3000 as dts overlay

Add overlay dts to use the qorvo-dw3000 with spidev.
The bootconf is named "mt7622-bananapi-bpi-r64-qorvo-dw3000".

Use fw_printenv to look which bootconf is selected.

Signed-off-by: Nick Hainke <vincent@systemli.org>

PolynomialDivision committed:
uboot-mediatek: bpi-r64: add bootconf for dw3000

Add dw3000 dts overlay which is using spidev.

Signed-off-by: Nick Hainke <vincent@systemli.org>

PolynomialDivision commented:
@dangowrt Like that?

dangowrt commented on target/linux/mediatek/patches-5.10/900-dts-mt7622-bpi-r64-aliases-for-dtoverlay.patch:
```diff
@@ -78,3 +78,38 @@
+ 
+ #ifdef CONFIG_OF
+ static const struct of_device_id spidev_dt_ids[] = {
++	{ .compatible = "bpi,mt65xx" },
```
Put the spi **client** device here. `bpi,mt65xx` is the spi **host**.

PolynomialDivision replied:
Okay, maybe I'm trying to do something really stupid. I have a DW3000. I have to communicate via spi with it. I thought I just use spidev for doing communication from userspace. There was a discussion that you need to add some compatible string to [spidev driver](http://forum.banana-pi.org/t/bpi-r2-spi-communication/4779/138). Now after reading the whole discussion I'm unsure if I'm still doing it right since the actual dev for bpi said that adding some compatibility string is wrong. Now you also say that I should add my spi client device, so the dw3000?

Actually, adding `compatible = "spidev";` in the node should be enough.

PolynomialDivision replied:
Okay, [stackoverflow](https://stackoverflow.com/a/53639989/8474618) also says I should just add my device as compatible string. Then I do it. ;)

PolynomialDivision replied:
Or I just use https://github.com/torvalds/linux/blob/master/drivers/spi/spidev.c#L678 ?

dangowrt replied:
I guess if that DW3000 hardware requires a driver using spi-dev in user-space, you should add "qorvo,dw3000" as compatible string.

dangowrt reviewed (commented):
We might need to change `include/image-commands.mk` and `scripts/mkits.sh` to allow generating FIT configs with a combination of DTS overlays, so that you could have your SPI mod **and** PCIe or SATA.
Right now we simply generate one config for each DT overlay because there wasn't yet a need to do anything more complex.

dangowrt reviewed (commented):
I guess if this is about "rohm,dh2228fv", then the name of the overlay should also be related to that "rohm,dh2228fv" name. spi0 is misleading, as spi is always available and doesn't depend on pinctrl choices (unlike pcie2 vs. sata, which does depend on pinctrl choice).

PolynomialDivision commented:
For me it just feels a bit like code duplication, if you just want to access the spi interface using spidev.

PolynomialDivision commented:
Changed everything accordingly and run-tested again on r64.

dangowrt commented on target/linux/mediatek/patches-5.10/900-dts-mt7622-bpi-r64-aliases-for-dtoverlay.patch:
```diff
@@ -78,3 +78,38 @@
+@@ -0,0 +1,22 @@
++/* SPDX-License-Identifier: (GPL-2.0-only OR MIT) */
++
++#include <dt-bindings/gpio/gpio.h>
```
You aren't using any symbols from `gpio.h`, are you?

PolynomialDivision replied:
nope, forgot to delete that.

dangowrt commented on target/linux/mediatek/image/mt7622.mk:
```diff
@@ -74,7 +74,7 @@ define Device/bananapi_bpi-r64
   DEVICE_MODEL := BPi-R64
   DEVICE_DTS := mt7622-bananapi-bpi-r64
-  DEVICE_DTS_OVERLAY := mt7622-bananapi-bpi-r64-pcie1 mt7622-bananapi-bpi-r64-sata
+  DEVICE_DTS_OVERLAY := mt7622-bananapi-bpi-r64-pcie1 mt7622-bananapi-bpi-r64-sata mt7622-bananapi-bpi-r64-qorvo-dw3000
```
We should find a better way to include multiple DTS overlays, not using bootconf but using U-Boot's Extension board management sub system instead. Because now we'd need 4 configs, pcie1, sata, pcie1+dw3000, sata+dw3000 which will multiply further for every new feature we add...

PolynomialDivision replied:
okay

PolynomialDivision replied:
I will have a look how I can do that.

dangowrt replied:
I found a good example of loading dynamic DT overlay here:
https://www.96boards.org/documentation/consumer/dragonboard/dragonboard410c/guides/dt-overlays.md.html#3-load-overlays-dynamically

As adding the spi device is not needed for early boot it should be possible to just load the DTB in Linux at runtime (SATA or NVME storage in PCIe slot for example could be used for extroot hence DT overlay needs to be applied early by U-Boot).

We should define a nice way to do that, ie. how to package dt-overlay packages in OpenWrt and have the overlay(s) applied at boot (in the right order).

PolynomialDivision replied:
@dangowrt Just saw you enabled: https://github.com/openwrt/openwrt/commit/401d7ebf2c9bdcc7c1e24900ed9f8acf87d932c8 :)

dangowrt replied:
Yes, and I had your case in mind.
Now you should be able to simply load dt overlay at runtime after mounting configfs and placing it there.

PolynomialDivision replied:
I also found your discussion here:
https://patchwork.ozlabs.org/project/openwrt/patch/YZ7hf0DVfM7LSobG@makrotopia.org/

However, it is not working for me. Did I miss anything?

PolynomialDivision replied:
Ah okay,
> However, if this whole approach is a dead-end and will just burden us
with additional maintainance in the long run, we have to find a better
way.

Sad :(
